### ISSUE TITLE
Flutter debugging in Visual Code throws handled exceptions as uncaught

### ISSUE THREAD

**Issue by cosinus84** [flutter/flutter#33427]
Runing in . Visual Code (debugging)  ignores try/catch. Handled exception is treated as uncaught exception.
https://github.com/flutterchina/dio/issues/314
Running the app with `flutter run` is working as expected or directly from XCode

[✓] Flutter (Channel dev, v1.6.3, on Mac OS X 10.14.4 18E226, locale en-RO)
    • Flutter version 1.6.3 at /Users/cosinus/flutter
    • Framework revision bc7bc94083 (5 days ago), 2019-05-23 10:29:07 -0700
    • Engine revision 8dc3a4cde2
    • Dart version 2.3.2 (build 2.3.2-dev.0.0 e3edfd36b2)

[✓] VS Code (version 1.34.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.0.2

The function calling updateIoaOn wraps in a try/catch(it throws a 405 Exeption code).
```
Future<bool> updateIoaOn(bool status) async {
    FormData formData = new FormData.from({"status": status ? 1 : 0});
    Response response = await dio
        .post(_appStore.getURL("test/ioa-on"), data: formData);
    return response.data["result"] == 1 ? true : false;
  }
```
On a tree of async->await functions (on a deep level).

**Comment by kangwang1988** [comment]
cc @DaveShuckerow

**Comment by cosinus84** [comment]
The error comes from visual code debugger, if you use flutter run from console then all works as it should. It happened to me with my own code and I posted an issue.

**Comment by DanTup** [comment]
@cosinus84 if you switch to the Debug side bar, which options are ticked at the bottom? If "All Exceptions" is ticked then the debugger will stop on all thrown exceptions:

<img width="365" alt="Screenshot 2019-06-05 at 1 05 04 pm" src="https://user-images.githubusercontent.com/1078012/58954985-9a95e300-8792-11e9-8d5a-7f8d33a0be17.png">

Normally you would only have "Uncaught exceptions" checked.

**Comment by cosinus84** [comment]
"All Exceptions" is not checked, I have a lot of try/catch blocks, it happened only on some deep tree async/await functions. I see now that some new updates, I will test with this ones and I will write back. 
Was not making any sense that all Dio(library) calls exceptions are treated corectly and one was failing to be treated right in visual code but was working with flutter run. 

The same is true for futter_cached_network_image, I was getting the HttpException but the lib is providing errorWidget. With flutter run all works ok.

**Comment by cosinus84** [comment]
![Screenshot 2019-06-06 at 00 30 44](https://user-images.githubusercontent.com/1459132/58992052-6225f180-87f2-11e9-96f0-f087e578383d.png)

**Comment by DanTup** [comment]
I wonder if this is the same as https://github.com/dart-lang/sdk/issues/36911 then. I thought it only occurred when using the VM flags, but maybe it's happening more than that.

Could you capture a log of this happening?

- Run the **Dart: Capture Logs** command from the VS Code command palette
- Untick everything except Debugger (leave that ticked)
- Reproduce the issue
- Click **Stop Logging** on the logging notification (which may have collapsed into the bell icon in the status bar)
- Attach the log

Thanks!

**Comment by cosinus84** [comment]
Thanks, here is the log
[Dart-Code-Log-2019-05-04 09-43-19.txt](https://github.com/flutter/flutter/files/3260302/Dart-Code-Log-2019-05-04.09-43-19.txt)

**Comment by DanTup** [comment]
@a-siva this looks like https://github.com/dart-lang/sdk/issues/36911. This time none of the VM flags were used (since it was launched by Flutter). We definitely set the mode to Unhandled in the log:

```
[09:44:13] ==> {"id":"889","method":"setExceptionPauseMode","params":{"isolateId":"isolates/226212136","mode":"Unhandled"}}
```

And then the VM paused on the exception shown in the screenshot above that is surrounded with `try`/`catch`:

```js
{
	"jsonrpc": "2.0",
	"method": "streamNotify",
	"params": {
		"streamId": "Debug",
		"event": {
			"type": "Event",
			"kind": "PauseException",
			"isolate": {
				"type": "@Isolate",
				"fixedId": true,
				"id": "isolates\/226212136",
				"name": "main",
				"number": "226212136"
			},
			"timestamp": 1559803463377,
			"topFrame": {
				"type": "Frame",
				"kind": "Regular",
				"location": {
					"type": "SourceLocation",
					"script": {
						"type": "@Script",
						"fixedId": true,
						"id": "libraries\/@131252778\/scripts\/package%3Adio%2Fsrc%2Fdio.dart\/16b2b87d2d2",
						"uri": "package:dio\/src\/dio.dart",
						"_kind": "kernel"
					},
					"tokenPos": 24366
				},
				"function": {
					"type": "@Function",
					"fixedId": true,
					"id": "classes\/1412\/closures\/491",
					"name": "async_op",
					"_vmName": ":async_op",
					"owner": {
						"type": "@Function",
						"fixedId": true,
						"id": "classes\/1412\/functions\/_makeRequest%40131252778",
						"name": "_makeRequest",
						"_vmName": "_makeRequest@131252778",
						"owner": {
							"type": "@Class",
							"fixedId": true,
							"id": "classes\/1412",
							"name": "Dio"
						},
						"_kind": "RegularFunction",
						"static": false,
						"const": false,
						"_intrinsic": false,
						"_native": false
					},
					"_kind": "ClosureFunction",
					"static": false,
					"const": false,
					"_intrinsic": false,
					"_native": false
				},
				"code": {
					"type": "@Code",
					"fixedId": true,
					"id": "code\/22b57aae-13261f1e0",
					"name": "[Unoptimized] Dio._makeRequest",
					"_vmName": "[Unoptimized] async_op",
					"kind": "Dart",
					"_optimized": false,
					"_intrinsic": false,
					"_native": false
				},
				"vars": [
					{
						"type": "BoundVariable",
						"name": "this",
						"value": {
							"type": "@Instance",
							"class": {
								"type": "@Class",
								"fixedId": true,
								"id": "classes\/1412",
								"name": "Dio"
							},
							"kind": "PlainInstance",
							"id": "objects\/0"
						},
						"declarationTokenPos": -1,
						"scopeStartTokenPos": 21964,
						"scopeEndTokenPos": 24473
					},
					{
						"type": "BoundVariable",
						"name": "cancelToken",
						"value": {
							"type": "@Instance",
							"_vmType": "null",
							"class": {
								"type": "@Class",
								"fixedId": true,
								"id": "classes\/138",
								"name": "Null"
							},
							"kind": "Null",
							"fixedId": true,
							"id": "objects\/null",
							"valueAsString": "null"
						},
						"declarationTokenPos": 21985,
						"scopeStartTokenPos": 21964,
						"scopeEndTokenPos": 24473
					},
					{
						"type": "BoundVariable",
						"name": "options",
						"value": {
							"type": "@Instance",
							"class": {
								"type": "@Class",
								"fixedId": true,
								"id": "classes\/1434"…
```

**Comment by ZichangG** [comment]
Not sure why it happens from the log. I can't reproduce locally. 
@cosinus84 Do you have a sample program for me to run?

**Comment by cosinus84** [comment]
@ZichangG I created an repo https://github.com/cosinus84/exctest

**Comment by ZichangG** [comment]
Thanks for example. I think this is relative to how VM determines whether exceptions are handled or not.
As I answered in #36911,
> VM will checks try catch block to determine an exception is handled or unhandled. The expectLater() is basically using Future.then() machinery instead of try catch block. So the exception thrown was considered as unhandled exception. The workaround is to use try catch block.
> 
> From our perspective, it is hard to trace Future objects and registered callbacks. We are still going to use trycatch block as criteria.

Inside dio.dart, it also propagates the error using Future.then.catcherror. Although 2 mechanisms are quite similar in terms of functionality,  the difference is frames on the stack. VM managed to track all frames on the stack including frames separated by async/await calls. But not for Future, it will wiped out all the previous frames on the stack. In other words, try catch block should be on one of those frames in order to determine exception is handled.

**Comment by DanTup** [comment]
@ZichangG in the analysis server example (from https://github.com/dart-lang/sdk/issues/36911), I don't think the original code is using `.then()` (although when I repro'd it in a test, that did use `expectLater`). It does use `runZoned`, I don't know if that has the same limitations.

This limitation is quite awkward - it means when running a test run with the debugger attached, you keep stopping at places that throw exceptions even though they're being caught and have to press Continue (or, you have to turn off break-on-unhandled-exceptions, but then if some tests fail, you have to go back and re-run them individually to debug).

That said, I just tried this - and the first code pauses on the exception, but the second one does not (even though it does not have a `try/catch`). Any ideas why these are different?

```dart
main() async {
  // Breaks on caught exception (unexpected).
  () async {
    throw 'err1';
  }()
      .catchError((e) => print(e));

// Does not break (expected).
  () async {}().then((a) => throw 'err2').catchError((e) => print(e));
}
```

**Comment by ZichangG** [comment]
The first case is just a closure throwing exception. When the exception was thrown, frames on the stack doesn't contain try catch blocks. So debugger paused on unhandled exception.

The second case has Future.then() involved. () async {}() returns a future object which has 2 listeners following (then and catchError). It called _propagateToListeners() to pass value/error. [handleValueCallback](https://github.com/dart-lang/sdk/blob/792a3d561a14ebc83ce527f2cb70c839b9909122/sdk/lib/async/future_impl.dart#L637) has a try catch block here. It then invoked closure that throws exception. Debugger at this time looked at stacktrace and found try/catch block in handleValueCallback() and treat it as handled exception.

I'm not fully understand how it works. I think the difference is position where exception is thrown. Second one throws inside then().

**Comment by DanTup** [comment]
Oh, I see. That still leaves me confused about this one then (apologies for jumping between them - we could close this as a dupe of that the other issue if we think they're the same):

![Screenshot 2019-05-09 at 3 18 01 pm](https://user-images.githubusercontent.com/1078012/57461011-212bd300-726e-11e9-8134-6821127b3fb9.png)

(screenshot is from https://github.com/dart-lang/sdk/issues/36911)

That one does have a `try`/`catch` around it, but was also breaking as an unhandled exception

**Comment by cosinus84** [comment]
As I see you guys seems to know a lot of "how things work" inside dart&flutter. Can you please say in a couple of words or take a conclusion about the issue so me and others understand what is all about. Thank you

**Comment by DanTup** [comment]
I'm not very familiar with the VM, but my understanding (which @ZichangG can hopefully confirm or correct) is that when an exception is thrown, the debugger needs to figure out whether it's going to be handled or not (so that it can support breaking only on unhandled exceptions). I think it does this by looking up the call stack for `try` blocks (and then probably looking at any conditions on their `catch`s?), but in the case where there is no catch block (eg. calling `catchError` on a future) it doesn't find this and therefore it considers the exception is unhandled.

I now wonder if this is why tests seem to break on unhandled exceptions even though the test framework is handling the error (which is good functionality, even though not normally what you'd want).

**Comment by komapeb** [comment]
@cosinus84 The workaround (for the time-being) is to disable your IDE's option to break on unhandled exceptions or just ignore the exception. Not optimal, I know... Here are some related issues:

- https://github.com/flutter/flutter/issues/26705
- https://github.com/dart-lang/sdk/issues/37268
- https://github.com/dart-lang/sdk/issues/36911

**Comment by ZichangG** [comment]
> when an exception is thrown, the debugger needs to figure out whether it's going to be handled or not (so that it can support breaking only on unhandled exceptions). I think it does this by looking up the call stack for `try` blocks
>
That's basically how we determine exceptions are handled or not. A small correction here, try blocks itself is not handler. It must be provided with catch statement(This is where exceptions are being handled). For instance, a try finally block without catch is not going to work.

>(and then probably looking at any conditions on their `catch`s?),
>
Yup, it also checks whether the exception thrown matches the expected type written in the catch block.

> but in the case where there is no catch block (eg. calling `catchError` on a future) it doesn't find this and therefore it considers the exception is unhandled.
>
This is exactly what happened. I think it is probably better to document this subtle difference on how to interpret "handled exceptions". 

I check docs for dart api page. It does use 'handled' for future.then() and future.catcherror(). Make sense to me, as users indeed provide a way to handle exceptions. But from VM perspective, we interpret the ''handled" exceptions differently. VM will check only call stacks to find handler. It is hard to trace future objects especially a callback can be registered after the exception was thrown.

"Uncaught exception" in VS code is not equivalent to "unhandled exception" that VM expected. Is there any place we can document to remove ambiguity? What do you think? @DanTup

**Comment by komapeb** [comment]
@ZichangG considering this, is it safe to say that, to avoid this behavior, we should be using `async` and `await` in combination with `try...catch` blocks, instead of `someFuture.then(...).catchError(...)`?

**Comment by ZichangG** [comment]
> @ZichangG considering this, is it safe to say that, to avoid this behavior, we should be using `async` and `await` in combination with `try...catch` blocks, instead of `someFuture.then(...).catchError(...)`?
>
Yes.

**Comment by jamesdixon** [comment]
I'm also running into this same issue while debugging in VSCode. However, my API call is in a `try`/`catch` block and not using `catchError()`.

```dart
  Future<dynamic> login({
    @required String email,
    @required String password,
  }) async {
    try {
      final data = {"email": email, "password": password};

      final response = await _http.post(
        'https://api.url/token',
        data: _encodeRequest(data),
      );

      return _extractData(response);
    } on DioError catch (e) {
      print(e.response.statusCode);
    }
  }
````

Instead of the error being caught, the debugger is pausing and I have to resume in order for it to trigger the `catch` block.

Am I missing something?

`flutter doctor` output:

```
[✓] Flutter (Channel stable, v1.7.8+hotfix.3, on Mac OS X 10.14.5 18F132, locale en-US)
    • Flutter version 1.7.8+hotfix.3 at /Users/jamesdixon/flutter
    • Framework revision b712a172f9 (9 days ago), 2019-07-09 13:14:38 -0700
    • Engine revision 54ad777fd2
    • Dart version 2.4.0
```

**Comment by feinstein** [comment]
I don't think that your library call matters here, the problem is the call inside the library itself, as the VM won't be able to understand the stack at that point and won't know to whom it should be delivered to.

**Comment by ZichangG** [comment]
> I don't think that your library call matters here, the problem is the call inside the library itself, as the VM won't be able to understand the stack at that point and won't know to whom it should be delivered to.

That's probably what happens. Your function is good, perhaps somewhere deeper in the library causes this problem.

**Comment by jamesdixon** [comment]
@feinstein @ZichangG thanks for the explanation!

**Comment by larssn** [comment]
cc @ZichangG @DanTup 
Is this supposedly fixed? If not, why was it closed?

We're seeing **handled** exceptions, thrown somewhere in app, get picked up by the error handler in `runZoned`; **even when the app is built for production**.

Example:
```
runZoned(() {
    runApp(MainWidget());
  }, onError: (error, stackTrace) {
    print('Global error handler!!');
    print(error);
});
```

The above will often pick up handled exceptions, which is a major problem as most global error handlers will report errors to an error logging service.

An actual example of this happening can be created using the plugin `flutter_cached_network_image`, and having it point to an image that does not exist. An internal exception is thrown in a Stream which is consumed by a StreamBuilder (which handles Stream errors); but the global error handler above still catches the error.

**Comment by ZichangG** [comment]
Not quite familiar with flutter. By looking at Hixie's reply from the issue you referenced,

> Actually fetching the asset is done asynchronously, when the widget is inflated, placed into the tree, and has its State object created. Creating the Image object itself doesn't do anything (creating widgets in general doesn't do anything, they're just immutable state descriptions). That's why this isn't catching the exception: it's not thrown until much late
>

Exception was thrown very late when original stacktrace has already changed. That's why we didn't find the exception handler at current stacktrace when exception is thrown. This exception will be thrown as unhandled exception.

We don't have a good way to track exception handlers except looking at stacktrace.

**Comment by lsegal** [comment]
Not entirely sure why this is marked as closed when the issue is still persisting for users. Debugging in VSCode (and probably others) is currently really hard with any app that relies on completely _valid_ usage of asynchronous try/catch in Dart especially when time-sensitive things are happening when the debugger chooses to handle what it thinks is an uncaught exception.

TLDR: the debugger is pausing on perfectly normal code and that should be considered a bug and fixed somewhere in Flutter's tooling.

Is there another issue this is being tracked in? I see a few related issues but unsure where to continue the conversation.

**Comment by feinstein** [comment]
In a nutshell, this is closed because there's nothing to be done, Dart can't know the Debug stacktrace in order to deliver the exception to the right catcher.

If you go back on this thread you will find several issues related to this, some in the Dart language repo, where the Dart team explains what's happening and why it can't be fixed.

**Comment by lsegal** [comment]
I guess the contention here is "it can't be fixed" isn't really good enough for the users reporting it. 

If this isn't a solvable problem (I can't believe that this is true given that it's already solved in most other languages), then that's a huge glaring omission in the tooling and commitment to support a debugger. The UX here is outright _broken_, especially when the standard practice for Flutter development is to do most of your work while the app is running. Having exceptions pop that you've explicitly caught can be at minimum frustrating and in most cases completely disabling us from using this flow.

What about an option to [temporarily] disable the catching of exceptions and letting them bubble past the debugger so we don't get bogged down by the tooling? Is that something Flutter has control over?

**Comment by feinstein** [comment]
I am at the street right now, so it's hard for me to get the issue number, but the right place to discuss this is not here, but on the issue at the Dart language repo. I am not aware of the specifics, I am just relaying what the Dart team said about the Dart VM, as this behavior doesn't manifest on release compiled code.

There are ways to achieve what you want in several different IDEs, go look the issue about the Firebase Auth problem and there someone showed how to disable this, like you want.

In order to avoid this problem, use async await, instead of Future.then.catch, I know this can't be done in libraries, but at least we can do it in our own code (almost always).

**Comment by larssn** [comment]
"Impossible" is usually a synonym of "It's too much work", in developer-speak.

**Comment by cosinus84** [comment]
In my opinion the sky had fallen on the dart team when they heard the decision that dart was the
winner for flutter. After years of meetings and talking as the Ents from the Lord of the Rings (to be or not to be semicolons, etc...) they started to work on dart with demonic speed by adding, fixing and removing. 
Flutter needs a modern language, dart is no so far away from that point. Flutter changed the peace and quiet of dart headquarter. As major changes were added from 1.0 to 2.0 they can do this with 3.0. 
I personally had big trouble with Dio library and this problem, I actually have the feeling of doing React Native (easily and steady with new versions screws begin to fail).

**Comment by lcsvcn** [comment]
2020 is almost ending and this issue still remains...

Will someday it be fixed? It is really bad to have handled errors pausing you workflow.

I am having this issue with FirebaseAuth.

**Comment by midnight-cigarettes** [comment]
Just had this exact same issue that wasted too much of my time.

I've been using `Image.network` with errorBuilder.
Even the sample code here throws an exception that appears as if it is "uncaught": https://api.flutter.dev/flutter/widgets/Image/errorBuilder.html
It throws an uncaught when it **shouldn't** because I've put a way to handle it.

The only way I was able to finally "solve" it is by unchecking uncaught exceptions out of desperation:
Which shouldn't really be that way...
![Screenshot 2020-12-24 131920](https://user-images.githubusercontent.com/65785136/103062916-a6523000-45ea-11eb-886d-0563ebdd784c.png)

I don't know exactly if this is a Flutter or a Dart issue, but if someone can take a look at this then that would be amazing.
As this is just a very misleading error that could lead to a rabbit hole of investigation that's not really a fault of the user.

For context, I'm using:
VSCode: 1.52.1
Flutter VSCode Extension: 3.17.0
Dart VSCode Extension: 3.17.1

Here's Flutter Doctor:
```
[√] Flutter (Channel stable, 1.22.5, on Microsoft Windows [Version 10.0.18363.1256], locale en-US)
    • Flutter version 1.22.5 at F:\Windows\flutter
    • Framework revision 7891006299 (13 days ago), 2020-12-10 11:54:40 -0800
    • Engine revision ae90085a84
    • Dart version 2.10.4

 
[√] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at F:\Windows\android-sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_SDK_ROOT = F:\Windows\android-sdk
    • Java binary at: F:\Windows\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[!] Android Studio (version 4.1.0)
    • Android Studio at F:\Windows\Program Files\Android\Android Studio
    X Flutter plugin not installed; this adds Flutter specific functionality.
    X Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[√] VS Code, 64-bit edition (version 1.52.1)
    • VS Code at C:\Program Files\Microsoft VS Code
    • Flutter extension version 3.17.0

[√] Connected device (1 available)
    • Android SDK built for x86 (mobile) • emulator-5554 • android-x86 • Android 10 (API 29) (emulator)

! Doctor found issues in 1 category.
```

**Comment by tamoyal** [comment]
not sure why this is closed, problem still exists

**Comment by DanTup** [comment]
I think many of the issues above are the same as https://github.com/dart-lang/sdk/issues/37953 which has been recently fixed in the Dart VM and should be included in a future Flutter release.

**Comment by kevinmarrec** [comment]
I've been testing the fix they did, it sounds that there's still an issue when there are multiple `Future` chained.

With last Dart `dev` version i'm still not able to correctly catch third parties exceptions (ex: `http`) without having VS Code telling me it's uncaught.

See last comments in https://github.com/dart-lang/sdk/issues/37953

Some further work is still needed on Dart SDK then.

**Comment by imedboumalek** [comment]
The problem persists, I still have it on 5 different machines (all running different linux distros), and have had it since flutter 1.22.6 (now I'm on flutter 2.2.3)
